### DISCUSSION TITLE
Infinite loop

### DISCUSSION THREAD

**Discussion by Anixx** [valinet/ExplorerPatcher#465] [General]
This is what happened to me two times already.

1. Crash of DWM. I think ExplorerPatcher is not the cause, but cannot be sure.

2. After the crash I am thrown into the logon screen. I re-login. DWM is not on (no acceleration of windows).

3. The taskbar appears briefly, but with no task buttons, black unskinned areas, no start menu and tray.

4. Then shortly after that the taskbar disappears and I see a message box that "An error had occured when servicing this software..."

5. The taskbar briefly re-appears again, and this continues in an infinite loop.

If I shut off the computer with the power button, the infinite loop continues after logon.

Last time I managed to get from the loop by clickling on the taskbar in a brief moment when it was visible and entering Explorer Patcher's preferences. There I unchecked the "Notify about updates" feature. After that the Explorer stopped restarting.

My guess is that one cause of the problem is that the Explorer Patcher tries to check for updates even when it just had re-started the Explorer due to failed check for updates. Possibly it should check for updates only once at logon.

The other cause of the problem is if the checking for updates failed, it restarts the Explorer anyway. Possibly, it should not.

**Comment by valinet** [comment]
Currently, it does not restart when the update check fails. Restarting is done by the updater anyway. Are you on a supported build? Please investigate thoughtfully, establish clearly what is happening, no one can work based off presumptions. Try with, without it as shell extension, debug it a bit. EP has nothing to do with DWM unless set as shell extension, in which case it may be loaded in there as well, idk... Also, DWM crashing should not throw you on the login screen, but only reload the image on the screen. `winlogon` crashing might throw you to the login screen. Anyway, try to reproduce everything on a clean install or to find a clear reproducible pattern.

Thanks

**Reply by Anixx** [reply]
I am on build 22000.318 currently, this is a non-insider branch.

> Also, DWM crashing should not throw you on the login screen, but only reload the image on the screen.

This happened 3 times with me already. I don't know, may be the video driver crashed. I fall into logon screen with a nature picture at background and when I re-login, the video acceleration is not working. Tearing when moving windows, etc. This all happened 3 times. The first time happened before the auto-check-for-updates feature was introduced, so there was no loop and no error message. The second time there was the loop, but I do not remember how did I manage to get out (likely, Ctrl+Alt+Del and then reboot). Third time I disabled auto-check-for-updates, and that got me out from the loop.

**Reply by valinet** [reply]
I don't see how check for updates would crash your video driver...?

**Reply by Anixx** [reply]
I did not claim so. The crash all times happened before the check for updates. But the check for updates locked me in the loop of restarting explorer.

**Reply by Anixx** [reply]
After logon, the message "An error had occured when servicing this software...", then explorer restart, then again the same message, again restart, and so on.

**Reply by valinet** [reply]
Are you certain it did so? Have you checked with the console enabled? If so, then attach a debugger (like WinDbg) to `explorer.exe` and see where it crashes. Or use [these instructions](https://docs.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps) to enable full crash dumps, reproduce it, then send me the crash dump so I can look on it.

**Reply by valinet** [reply]
> After logon, the message "An error had occured when servicing this software...", then explorer restart, then again the same message, again restart, and so on.

Set it to notify for updates if you have problems, not prompt to install, as you have UAC turned off and it automatically tries to install...

**Reply by valinet** [reply]
And I told you already, when using as shell extension, you cannot do too many updates without rebooting realistically, as the previous files are still in use in basically most of the processes already open on the system.

**Reply by Anixx** [reply]
I think, it did not crash (after logon). What I think, happened is 

1. It checked for updates.

2. For some reason EP killed explorer and started update, even though the option was on only to check for updates, not to update.

3. The EP's files were locked by some software (vault service or whatever). EP reports update error.

3. EP restarts explorer and checks for updates again.

**Reply by Anixx** [reply]
> you have UAC turned off and it automatically tries to install...

Yes! UAC was off.

**Reply by valinet** [reply]
Yeah, idk, you have to investigate it. Maybe it's easier if you enable UAC temporarily, as that slows things down so you can see if you set it to notify and it still tries to install automatically. I mean, it shouldn't be the case, it never does for me, the code seems to never have it do it...

**Reply by Anixx** [reply]
And I think it should not check for updates instantly after an unsuccessful update attempt and Explorer restart. I thought it would check for updates only on logon, once in a session.

**Reply by Anixx** [reply]
BTW, I keep the UAC off only due to the ugly immersive dialog... On Win 8.1 I liked the feature...

**Reply by valinet** [reply]
Yeah, idk, I very much prefer it like this, logging off to update would be very annoying... this combination of shell extension + UAC off + automatically install is niche... it will be addressed. I will add a question before starting applying an update when UAC is off, similar to what is shown in UAC, the reason being that without one, someone could hijack the registry entries and have this execute arbitrary binaries. That's why the UAC prompt shows the URL the file was downloaded for, so the user knows where it came from and can decide whether to run it or not. Some similar question has to be answered with UAC disabled as well. It's on the roadmap...

**Reply by valinet** [reply]
> BTW, I keep the UAC off only due to the ugly immersive dialog... On Win 8.1 I liked the feature...

Disabling UAC just for the looks of a dialog is very bad in my opinion. UAC is a very important security component. Without UAC and with an admin account, it's just a ticking time bomb, every process can do pretty much whatever it wants on the system. Without UAC and with a standard user, it's kind of a pain in the butt to use. Please enable UAC.

**Reply by Anixx** [reply]
> I very much prefer it like this, logging off to update would be very annoying... 

I meant, automatic check for updates should be only on logon, or at least, not after a failed update attempt. Manual check (by pressing the button) should be always available...

**Reply by Anixx** [reply]
> I will add a question before starting applying an update when UAC is off, similar to what is shown in UAC

This is a very good solution.

**Reply by Anixx** [reply]
> Please enable UAC.

OK. You had convinced me. But I (and not only me) am still looking for ways to restore the classic UAC.

**Reply by valinet** [reply]
Yeah, I understand, I mean, me too, like, I just looked, but in the mean time, why risk it... I guess replacing `consent.exe` with some version where `XamlCredUIAvailable` is not working, right? The problem is it's thrown in `System32` where it is messed up with all the system libraries. Maybe setting a "Debugger" for `consent.exe` in "Image Execution Options" in the registry (same way one redirects `taskmgr.exe` to some other executable) works and lets it load from some other folder where maybe all its dependencies could be stored alongside it and allow it to load...? I mean, there's still COM components being loaded from all over the place, but idk, maybe it works...

**Reply by valinet** [reply]
> > I will add a question before starting applying an update when UAC is off, similar to what is shown in UAC
> 
> This is a very good solution.

Thx. Btw, I don't know how to implement it, idk how to get the UAC level programmatically. I am sure there must be a trick or something, I just don't know it. I mean, I will figure it out eventually.

**Reply by Anixx** [reply]
> I guess replacing consent.exe with some version where XamlCredUIAvailable is not working, right?

I did not try this, but the guys who tried said it did not work for them.

> Maybe setting a "Debugger" for consent.exe in "Image Execution Options"

I had this idea. But you know, when a program is started as debugger, some debug command line options are sent to it. It does not matter for taskman.exe but consent.exe possibly has some command line arguments.

Also, it could be the case that the actual dialog is in some dll, not in the executable.

**Reply by Anixx** [reply]
> Btw, I don't know how to implement it, idk how to get the UAC level programmatically

Just check for updates and ask if one wants to update? If not, just do nothing?

**Reply by valinet** [reply]
> I had this idea. But you know, when a program is started as debugger, some debug command line options are sent to it. It does not matter for taskman.exe but consent.exe possibly has some command line arguments.

That can be confirmed by looking on the disassembly, or just by going for it and seeing what it does...

> Also, it could be the case that the actual dialog is in some dll, not in the executable.

So what? By having it in some separate folder, you can put its DLL dependencies beside it and those will be loaded first. As I said, this does not bypass COM components, which will get loaded using the information in the registry, but it might work.

Anyway, the key is to indeed, identify where the dialog originates from, how is it called, maybe that allows setting its type, who knows, if not, I am pretty sure the old dialog is still in there, it is just neutered by some `if` check probably and the `XamlCredUIAvailable` is never took into consideration now. I won't be surprised if it all boils down to editing 2 bytes or so in some weird DLL, it happened before =))) (with enabling the network and battery flyouts in Windows 11 which I did recently). The key to everything is studying the actual disassembly, it along with the symbols offer pretty solid information about how things work.

**Reply by valinet** [reply]
> > Btw, I don't know how to implement it, idk how to get the UAC level programmatically
> 
> Just check for updates and ask if one wants to update? If not, just do nothing?

???

Yeah, as I said, with UAC off, it will ask a question before applying the update. At the moment, I don't know how to determine if UAC is off or not though. Idk, I will find a way eventually, I am looking into it, that's all I was saying.

**Reply by Anixx** [reply]
Just checked. consent.exe is only 192 KB. Definitely, the dialog is not there.

But... Well, it could be just 192 KB...

**Reply by Anixx** [reply]
This is how task manager gets run as a debugger:

"C:\Program Files\ClassicTaskmgr\tm.exe" "C:\Windows\System32\Taskmgr.exe" /2

The name of the original file is sent to the debugger as an argument. (I used Process Hacker to see the command line)

**Reply by valinet** [reply]
> Just checked. consent.exe is only 192 KB. Definitely, the dialog is not there.

I like how you clearly invested the time to investigate it throughfully by disassembling it and taking its guts apart and you can clearly state that. No shade of doubt, indeed. 😂

Idk, you’re probably right, but being 192KB only is not an argument at all, 192KB can fit a LOT of stuff, especially a call to a function which takes as argument the type of consent dialog to create, for example, using internal or external code 😉

**Reply by Anixx** [reply]
There is also a file ConsentExperienceCommon.dll

**Reply by Anixx** [reply]
I actually did program in assembler and debug code when at school. On a Soviet analog of PDP-11. For, instance, fixed the game Joe Blade and some others so that it to load its files from a floppy using standard system call rather than from tape recorder.

I also wrote some addons the the OS graphical file manager that allowed to change the volume's (FAT12) label and some other properties via a graphical dialog displayed over the NC-like file manager. All in asm debugger (+using non-commands "WORD XXXXX" to call system interrupts).

I lost the interest in assembler though after transition to x86, because of the ugliness of its architecture.

**Reply by valinet** [reply]
Oh, wow, interesting. I mean, nowadays, tools like IDA come with decompilers, they don’t show you just the op codes and disassembly, they can also generate a C-like pseudocode based off it. Idk, try it, look on the system files, they are also augmented with symbols, and maybe you find some answers, they have to be there. That’s the only sure way to figure it out eventually, like, everything has to boil down, at some point, to a bunch of instructions executed on the CPU.

**Reply by Anixx** [reply]
Well, I tried to replace consent.exe (+mui) and it did not work, reporting 'filesystem error".

Replacing via debugger key just does not affect anything.
